seed/seed20_test.go

```diff
@@ -3507,3 +3508,208 @@ func (s *seed20Suite) TestPreseedCapableSeedAlternateAuthority(c *C) {
+	compareDirs(c, filepath.Join(s.SeedDir, "systems", srcLabel), destSystemDir)
+
+	err = copier.Copy(destSeedDir, copiedLabel, s.perfTimings)
+	c.Assert(err, ErrorMatches, fmt.Sprintf(`cannot create system: system %q already exists at %q`, copiedLabel, destSystemDir))
+}
+
+func (s *seed20Suite) TestCopyCleanup(c *C) {
+	s.makeSnap(c, "snapd", "")
+	s.makeSnap(c, "core20", "")
+	s.makeSnap(c, "pc-kernel=20", "")
+	s.makeSnap(c, "pc=20", "")
+	requiredFn := s.makeLocalSnap(c, "required20")
+
+	const label = "20191030"
+	s.MakeSeed(c, label, "my-brand", "my-model", map[string]interface{}{
+		"display-name": "my model",
+		"architecture": "amd64",
+		"base":         "core20",
+		"grade":        "dangerous",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              s.AssertedSnapID("pc-kernel"),
+				"type":            "kernel",
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name":            "pc",
+				"id":              s.AssertedSnapID("pc"),
+				"type":            "gadget",
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name": "required20",
+				"id":   s.AssertedSnapID("required20"),
+			}},
+	}, []*seedwriter.OptionsSnap{
+		{Path: requiredFn},
+	})
+
+	seed20, err := seed.Open(s.SeedDir, label)
+	c.Assert(err, IsNil)
+
+	err = seed20.LoadAssertions(s.db, s.commitTo)
+	c.Assert(err, IsNil)
+
+	copier, ok := seed20.(seed.Copier)
+	c.Assert(ok, Equals, true)
+
+	// remove a snap from the original seed to make the copy fail
+	err = os.Remove(filepath.Join(s.SeedDir, "snaps", "snapd_1.snap"))
+	c.Assert(err, IsNil)
+
+	destSeedDir := c.MkDir()
+	err = copier.Copy(destSeedDir, label, s.perfTimings)
+	c.Check(err, NotNil)
+
+	// seed destination should have been cleaned up
+	c.Check(filepath.Join(destSeedDir, "systems", label), testutil.FileAbsent)
+}
+
+func checkDirContents(c *C, dir string, expected []string) {
+	sort.Strings(expected)
+
+	entries, err := os.ReadDir(dir)
+	c.Assert(err, IsNil)
+
+	found := make([]string, 0, len(entries))
+	for _, e := range entries {
+		found = append(found, e.Name())
+	}
+
+	c.Check(found, DeepEquals, expected)
+}
+
+func compareDirs(c *C, expected, got string) {
+	expected, err := filepath.Abs(expected)
+	c.Assert(err, IsNil)
+
+	got, err = filepath.Abs(got)
+	c.Assert(err, IsNil)
+
+	expectedCount := 0
+	err = filepath.WalkDir(expected, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		expectedCount++
+
+		gotPath := filepath.Join(got, strings.TrimPrefix(path, expected))
+
+		if d.IsDir() {
+			c.Check(osutil.IsDirectory(gotPath), Equals, true)
+			return nil
+		}
+
+		c.Check(gotPath, testutil.FileEquals, testutil.FileContentRef(path))
+
+		return nil
+	})
+	c.Assert(err, IsNil)
+
+	gotCount := 0
+	err = filepath.WalkDir(expected, func(_ string, _ fs.DirEntry, err error) error {
```

Comment:
Maybe I am getting this wrong, but shouldn't this run through `got` folder instead?

Reply:
Good catch, thank you!